example/spec/distributionAPI/s3credentialsSpec.js

```diff
@@ -0,0 +1,87 @@
+'use strict';
+
+const { URL } = require('url');
+const got = require('got');
+
+const { models: { AccessToken } } = require('@cumulus/api');
+const { serveDistributionApi } = require('@cumulus/api/bin/serve');
+const {
+  EarthdataLogin: { getEarthdataAccessToken }
+} = require('@cumulus/integration-tests');
+
+const { loadConfig } = require('../helpers/testUtils');
+const {
+  setDistributionApiEnvVars,
+  stopDistributionApi
+} = require('../helpers/apiUtils');
+
+const config = loadConfig();
+
+
+/**
+ * Login with Earthdata and get response for redirect back to
+ * distribution API.
+ */
+async function getTestAccessToken() {
+  const accessTokenResponse = await getEarthdataAccessToken({
+    redirectUri: process.env.DISTRIBUTION_REDIRECT_ENDPOINT,
+    requestOrigin: process.env.DISTRIBUTION_ENDPOINT
+  });
+  return accessTokenResponse.accessToken;
+}
+
+
+describe('Distribution API', () => {
+  let server;
+
+  process.env.AccessTokensTable = `${config.stackName}-AccessTokensTable`;
+  const accessTokensModel = new AccessToken();
+
+  beforeAll(async (done) => {
+    setDistributionApiEnvVars();
+
+    // Use done() callback to signal end of beforeAll() after the
+    // distribution API has started up.
+    server = await serveDistributionApi(config.stackName, done);
+  });
+
+  afterAll(async (done) => {
+    stopDistributionApi(server, done);
+  });
+
+  describe('handles requests for temporary credentials', () => {
+    let accessToken;
+
+    afterAll(async () => {
+      await accessTokensModel.delete({ accessToken });
+    });
+
+    it('redirecting to Earthdata login for unauthorized requests to /s3credentials endpoint.', async () => {
+      const response = await got(
+        `${process.env.DISTRIBUTION_ENDPOINT}/s3credentials`,
+        { followRedirect: false }
+      );
+      const authorizeUrl = new URL(response.headers.location);
+      expect(authorizeUrl.origin).toEqual(process.env.EARTHDATA_BASE_URL);
+      expect(authorizeUrl.pathname).toEqual('/oauth/authorize');
+    });
+
+    it('returning Credentials for authorized requests to /s3credentials endpoint', async () => {
+      accessToken = await getTestAccessToken();
+      const response = await got(
```

Comment:
It seems like there _might_ be an opportunity for a small re-factor. I previously wrote a small helper for requesting files on the distribution API: https://github.com/nasa/cumulus/blob/master/packages/integration-tests/api/distribution.js#L16

It seems like the part of `getDistributionApiS3SignedUrl` actually setting the cookie header and doing the request could be pulled out into another function and used here? I'm not sure if that buys us anything other than consistency, so up to you if we really need to do it

Reply:
0cda0ce84